### PR TITLE
[#1381] Word* types to Natural, not Integer

### PR DESCRIPTION
**Pull request by jiegillet**
As described in #1381.

I'd be happy to add some unit tests, but I'd need some help ^^

This is my first PR here, let me know if I did something strange.

### CODE REVIEW

**Comment by sjakobi** [comment]
Cheers! I'd merge this as-is because the correctness seems obvious, but if you want to, you can add tests.

You could extend the unit tests, re-using the following helper:

https://github.com/dhall-lang/dhall-haskell/blob/6a8e7e12cc95376dfd9ddfba673e7d0ae0d3b191/dhall/tests/Dhall/Test/Dhall.hs#L332-L337

Doctests might be an even better fit. See e.g.

https://github.com/dhall-lang/dhall-haskell/blob/6a8e7e12cc95376dfd9ddfba673e7d0ae0d3b191/dhall/src/Dhall.hs#L359-L362

I run them with `stack test dhall:doctest` which is super slow for reasons that I'm afraid I don't understand.

**Comment by jiegillet** [comment]
When I run `stack test dhall:doctest`, I get:

```
/Users/jie/Documents/Programming/Haskell/dhall-haskell/dhall/ghc-src/Dhall/Crypto.hs:11:1: error:
    Ambiguous module name ‘Crypto.Hash’:
      it was found in multiple packages: cryptonite-0.25 hashing-0.1.0.1
   |
11 | import Crypto.Hash (SHA256)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by sjakobi** [comment]
Ah, that's unfortunate. Seems like `dhall-nix` has a transitive dependency on `hashing`. We don't depend on it directly though, so it really shouldn't be in scope. I think this might be worth reporting on the `stack` issue tracker.

Can you get rid of the error by running `stack clean` or `stack clean --force`?

**Comment by jiegillet** [comment]
> Can you get rid of the error by running `stack clean` or `stack clean --force`?

Neither worked.

**Comment by sjakobi** [comment]
How annoying! :(

Please use [package imports](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#package-qualified-imports) for `Crypto.Hash` then, and add comments why this is necessary.

**Comment by jiegillet** [comment]
> Can you get rid of the error by running `stack clean` or `stack clean --force`?

`stack clean --full` worked. 
Sorry about the previous message, it didn't work because I messed with the code and didn't clean up right.

**Comment by jiegillet** [comment]
New commit.

* I feel silly for adding such simple tests, but worth it for learning how to use dockets, I guess?
* Building the dockets failed again after I ran `stack build`, so I added `PackageImports`. Should I have committed that?

File: dhall/src/Dhall.hs [hunk 1]
```diff
@@ -1582,33 +1582,54 @@ instance Inject Int where
...
+>>> embed inject (12 :: Word)
+NaturalLit 12
+>>> embed inject (12 :: Word8)
+NaturalLit 12
+>>> embed inject (12 :: Word16)
+NaturalLit 12
+>>> embed inject (12 :: Word32)
+NaturalLit 12
+>>> embed inject (12 :: Word64)
+NaturalLit 12
```
Comment (sjakobi) [inline dhall/src/Dhall.hs]: Can you attach these doctests to the respective instance?

File: dhall/ghc-src/Dhall/Crypto.hs [hunk 2]
```diff
@@ -8,13 +9,13 @@ module Dhall.Crypto (
     ) where
 
 import Control.DeepSeq (NFData)
-import Crypto.Hash (SHA256)
+import "cryptonite" Crypto.Hash (SHA256) -- To avoid conflict with "hashing"
```
Comment (sjakobi) [inline dhall/ghc-src/Dhall/Crypto.hs]: That's a bit unclear – we don't directly depend on `hashing` after all.

Comment (Gabriel439) [reply dhall/ghc-src/Dhall/Crypto.hs]: Is the use of `PackageImports` still necessary now that the `stack` issue was resolved?  My understanding is that this disambiguation should not be necessary

Comment (jiegillet) [reply dhall/ghc-src/Dhall/Crypto.hs]: I realized I may not have the most recent version `stack`. I'll check when I get home.

**Comment by jiegillet** [comment]
Turns out after all, my version of `stack` was way behind, with the latest version, there are no issue. Apologies.

File: dhall/src/Dhall.hs [hunk 3]
```diff
@@ -1582,33 +1582,69 @@ instance Inject Int where
...
+
+instance Inject Word where
+    injectWith _ = InputType {..}
+      where
+        embed = NaturalLit . fromIntegral
+
+        declared = Natural
+
+{-| Injecting `Word8` types gives `Natural`
+>>> import Data.Word (Word8)
```
Comment (sjakobi) [inline dhall/src/Dhall.hs]: Can you move these imports to a setup section like this one:

https://github.com/dhall-lang/dhall-haskell/blob/bf40ec52a1d33bf2688d83fdee6bb334dd706cf8/dhall/src/Dhall/Map.hs#L677-L683

File: dhall/src/Dhall.hs [hunk 4]
```diff
@@ -1582,33 +1582,69 @@ instance Inject Int where
 
         declared = Integer
 
+{-| Injecting `Word` types gives `Natural`
+>>> embed inject (12 :: Word)
```
Comment (sjakobi) [inline dhall/src/Dhall.hs]: Please add an empty line after the first line of each haddocks section. I'm pretty sure haddock needs that.

File: dhall/src/Dhall.hs (hunk 3 above)
Comment (jiegillet) [reply dhall/src/Dhall.hs]: Where? Just before the instances? At the very top?

Comment (sjakobi) [reply dhall/src/Dhall.hs]: So far we've put these at the end of the files, which seems fine to me.

**Pull request merged**